imf/imf.py

```diff
@@ -488,15 +489,93 @@ def inverse_imf(p,
     else:
         raise NotImplementedError
 
+####This section contains the functions required to optimally sample a cluster####
+
+def _prefactor(max_star,massfunc):
+    """
+    Returns the multiplier required for an IMF to have at most one star above m_max.
+    """
+    return 1/massfunc.integrate(max_star,massfunc.mmax)[0]
+
+def _M_cluster(m,massfunc):
+    """
+    Returns the mass of a cluster distributed according to some IMF where the 
+    largest star has mass m.
+    """
+    k = _prefactor(m,massfunc)
+    return k*massfunc.m_integrate(massfunc.mmin,m)[0]+m
+
+def _max_star(m,M_res,massfunc):
+    """
+    Returns the most massive star capable of forming in a cluster of mass M_res
+    according to the m_max/M_cluster relation. Formatted for use with root finding.
+    """
+    return M_res-_M_cluster(m,massfunc)
+
+def _max_star_prime(m,M_res,massfunc):
+    """
+    Returns the derivative of _max_star at mass m. Used for Newton's method in
+    the case of an infinite upper bound on the provided mass function.
+    """
+    term1 = _prefactor(m,massfunc)**2*massfunc(m)*massfunc.m_integrate(massfunc.mmin,m)[0]
+    term2 = m*massfunc(m)*_prefactor(m,massfunc)
+    return -term1-term2-1
+
+def _get_next_m(m,last_m,k,massfunc):
+    """
+    Returns the next smallest star in an optimally sampled cluster given the 
+    previous star and overall IMF. Formatted for use with root finding.
+    """
+    return k*massfunc.m_integrate(m,last_m)[0]-m
+
+def _opt_sample(M_res,massfunc,tolerance):
+    """
+    Returns a numpy array containing stellar masses that optimally sample 
+    from a provided MassFunction to make a cluster with mass M_res.
+    """
+    #retrieve mass bounds from provided massfunc
+    mmin = massfunc.mmin; mmax = massfunc.mmax
```

Comment:
Most style guides require that we not use statements joined by semicolons like this
```suggestion
    mmin = massfunc.mmin
    mmax = massfunc.mmax
```